File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
     void reverse(vector<int>& nums , int start , int end){
        while(start<=end){
            int temp = nums[start];
            nums[start]=nums[end];
            nums[end]=temp ; 
            start++;
            end--;
        }
    }
    void rotate(vector<int>& nums, int k) {
       int size = nums.size();
       k = k % size ; 
       if(size==1) return ; 

       reverse(nums , 0 , size-1); //reverse whole array 

       reverse(nums , 0 , k-1);//reverse first k element 

       reverse(nums , k , size-1); //reverse next element till last element
    }
};

File: 0189-rotate-array/0189-rotate-array.cs
public class Solution {
    public void ReverseArray(int[] nums ,int start , int end){
        int i = start , j = end ; 
        while(i < j ){
            int temp = nums[i];
            nums[i]=nums[j];
            nums[j] = temp ; 
            i++;
            j--;
        }
    }
    public void Rotate(int[] nums, int k) {
        k = k % nums.Length ;
        if(nums.Length == 1) return ;
        ReverseArray(nums , 0 , nums.Length-1);
        ReverseArray(nums , 0 , k-1);
        ReverseArray(nums , k , nums.Length-1);

    }
}